### PR TITLE
Sum over dimensions

### PR DESCRIPTION
**Pull request by gordoncaleb**
Extends SumVertex to allow summing over specific dimensions.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 1]
```diff
@@ -386,10 +386,10 @@ public DualNumber log() {
         }
     }
 
-    public DualNumber sum() {
-        DoubleTensor sumOfAll = DoubleTensor.scalar(value.sum());
-        int[] resultDims = TensorShape.dimensionRange(0, value.getRank());
-        return new DualNumber(sumOfAll, this.partialDerivatives.sumOverOfDimensions(resultDims));
+    public DualNumber sum(int[] sumOverDimensions) {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Makes sense

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java [hunk 2]
```diff
@@ -64,17 +61,22 @@ public void putWithRespectTo(VertexId id, DoubleTensor value) {
      * This will sum partial derivatives that are represented as tensors over given dimensions.
      * The dimensions that are summed over will be reshaped to a scalar shape of 1x1.
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: Good catch. I've updated it.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java [hunk 3]
```diff
@@ -64,17 +61,22 @@ public void putWithRespectTo(VertexId id, DoubleTensor value) {
...
      *
-     * @param ofDimensions dimensions to sum over (should be first n dimensions)
+     * @param sumOverDimensions dimensions to sum over
+     * @param resultShape       shape of sum result
+     * @param sumOfRank         the rank of the of part of the partials
      * @return summed and reshaped partials
      */
-    public PartialDerivatives sumOverOfDimensions(int... ofDimensions) {
+    public PartialDerivatives sumOverOfDimensions(int[] sumOverDimensions, int[] resultShape, int sumOfRank) {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: Good call.

[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivatives.java]: I was going to hold off on refactoring the other sumOverWrtDimensions for the fixing broadcasting work I've also got in flight. I've gone ahead and made it match the other summation just for consistency. I expect in the next refactor or two there will only be one sum that works for all cases.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 4]
```diff
@@ -2,42 +2,104 @@
...
+
+        return Ints.toArray(inputShapeList);
+    }
+
     /**
      * Performs a sum across each value stored in a vertex
      *
      * @param inputVertex the vertex to have its values summed
      */
     public SumVertex(DoubleVertex inputVertex) {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: 👍

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 5]
```diff
@@ -2,42 +2,104 @@
...
+            reshapedDiffWrtSelf.putWithRespectTo(partialDerivative.getKey(), reshapedPartialDerivative);
+        }
+
+        PartialDerivatives derivativesWrtInput = reshapedDiffWrtSelf
+            .multiplyAlongWrtDimensions(DoubleTensor.ones(inputVertex.getShape()), wrtShape);
 
         return singletonMap(inputVertex, derivativesWrtInput);
     }
+
+    private int[] summedOverShape(int[] shape, int[] sumOverDimensions) {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: 👍

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java [hunk 6]
```diff
@@ -210,7 +210,7 @@ public void cannotSetIfMaskLengthIsSmallerThanTensorLength() {
     @Test
     public void cannotSetIfMaskLengthIsLargerThanTensorLength() {
         DoubleTensor tensor = Nd4jDoubleTensor.scalar(3);
-        DoubleTensor mask = Nd4jDoubleTensor.create(new double[] {1., 1., 1., 1.}, new int[] {2, 2});
+        DoubleTensor mask = Nd4jDoubleTensor.create(new double[]{1., 1., 1., 1.}, new int[]{2, 2});
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: .ones() (and we could fix that function to take a int... too?

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java [hunk 7]
```diff
@@ -239,6 +239,15 @@ public void canApplySqrt() {
...
     public void canBroadcastMultiplyRank4ContainingVectorAndMatrix() {
+
+        /*
+          a = np.array([1, 2, 3, 4, 5, 6, 7, 8]).reshape(2,2,2,1)
+          b = np.array([1, 2, 3, 4]).reshape(2,2,1,1)
+          ab = a * b
+          print(ab)
+          print(np.shape(ab))
+         */
+
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: Awesome idea.  Is there a comment anywhere making it clear if people aren't goint to immediately recognise this as NumPy stuff?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java [hunk 8]
```diff
@@ -28,7 +31,46 @@ public void doesSum() {
...
+        DoubleVertex a = new UniformVertex(shape, 0, 10);
+        DoubleTensor highrank = DoubleTensor.arange(0, TensorShape.getLength(shape)).reshape(shape);
+        a.setValue(highrank);
+
+        assertArrayEquals(new int[]{3, 5, 6, 1}, a.sum(0, 2).getShape());
+        assertArrayEquals(new int[]{5, 6}, a.sum(0, 1, 2, 5).getShape());
+        assertArrayEquals(new int[]{6, 1}, a.sum(0, 1, 2, 3).getShape());
+        assertArrayEquals(new int[]{2, 1}, a.sum(1, 2, 3, 4, 5).getShape());
+        assertArrayEquals(new int[]{1, 3}, a.sum(0, 2, 3, 4, 5).getShape());
+        assertArrayEquals(new int[]{1, 1}, a.sum(0, 1, 2, 3, 4, 5).getShape());
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java]: Should we have a test here for things that start and result in a scalar?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java [hunk 9]
```diff
@@ -118,21 +194,38 @@ public void canDoSumAutoDiffWhenWrtIsScalar() {
...
 
         finiteDifferenceMatchesForwardAndReverseModeGradient(ImmutableList.of(inputVertex), outputVertex, 1e-6, 1e-10);
     }
 
+    @Test
+    public void changesMatchGradientWhenSummingSpecificDimensions() {
+        DoubleVertex inputVertex = new UniformVertex(new int[]{2, 2, 2}, -10.0, 10.0);
+        inputVertex.setValue(DoubleTensor.arange(0, 8).reshape(2, 2, 2));
+
+        DoubleVertex outputVertex = inputVertex.sum(0).reshape(2, 2)
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java]: Should we need this reshape?  Shouldn't that just be the correct shape anyway?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 10]
```diff
@@ -2,42 +2,110 @@
...
-        PartialDerivatives derivativesWrtInput = derivativeOfOutputsWithRespectToSelf
-            .multiplyAlongWrtDimensions(DoubleTensor.ones(inputVertex.getShape()), this.getShape());
+        int[] wrtShape = summedOverShapeWithoutRankLoss(inputVertex.getShape(), overDimensions);
+
+        PartialDerivatives reshapedDiffWrtSelf = new PartialDerivatives(new HashMap<>());
+        for (Map.Entry<VertexId, DoubleTensor> partialDerivative : derivativeOfOutputsWithRespectToSelf.asMap().entrySet()) {
+            DoubleTensor partial = partialDerivative.getValue();
+
+            int[] newPartialShape = TensorShape.concat(
+                TensorShape.selectDimensions(0, partial.getRank() - getShape().length, partial.getShape()),
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: can we pull this apart a bit to make it clear what all the bits mean?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java [hunk 11]
```diff
@@ -2,42 +2,110 @@
...
+                wrtShape
+            );
+
+            DoubleTensor reshapedPartialDerivative = partialDerivative.getValue().reshape(newPartialShape);
```
Comment (christophernorth) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: won't this leave weird ones in the shape?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 12]
```diff
@@ -386,10 +386,10 @@ public DualNumber log() {
         }
     }
 
-    public DualNumber sum() {
-        DoubleTensor sumOfAll = DoubleTensor.scalar(value.sum());
-        int[] resultDims = TensorShape.dimensionRange(0, value.getRank());
-        return new DualNumber(sumOfAll, this.partialDerivatives.sumOverOfDimensions(resultDims));
+    public DualNumber sum(int[] overDimensions) {
+        DoubleTensor sumOfAll = value.sum(overDimensions);
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Excellent catch!

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java (hunk 6 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: Good catch.

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java (hunk 7 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: I meant to have these changes go in on a different PR where it was better explained. @GeorgeNash is looking at further Numpy checking. I'll add a little line on each of them saying it's numpy.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java [hunk 13]
```diff
@@ -59,12 +55,11 @@ public static INDArray divi(INDArray left, INDArray right) {
 
     private static INDArray broadcastDivide(INDArray a, INDArray b) {
         if (a.shape().length < b.shape().length) {
-            return broadcastMultiply(b, a.rdiv(1.0));
+            return broadcastMultiply(b.rdiv(1.0), a);
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java]: 👍

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java (hunk 8 above)
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertexTest.java]: 👍

File: keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java (hunk 7 above)
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensorTest.java]: 👍

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java (hunk 10 above)
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java]: 👍

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 14]
```diff
@@ -386,10 +386,10 @@ public DualNumber log() {
         }
     }
 
-    public DualNumber sum() {
-        DoubleTensor sumOfAll = DoubleTensor.scalar(value.sum());
-        int[] resultDims = TensorShape.dimensionRange(0, value.getRank());
-        return new DualNumber(sumOfAll, this.partialDerivatives.sumOverOfDimensions(resultDims));
+    public DualNumber sum(int[] overDimensions) {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: The dual number diffs?

[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Sorry, yeah.

[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: I'll deal with the merge conflicts on that end. I didn't want to make this PR any more complicated than it already was.

**Pull request merged**